opentsdb/opentsdb.go

```diff
@@ -0,0 +1,159 @@
+package opentsdb
+
+import (
+	"bufio"
+	"log"
+	"net"
+	"net/textproto"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/influxdb/influxdb"
+)
+
+const (
+	// DefaultPort represents the default OpenTSDB port.
+	DefaultPort = 4242
+
+	// DefaultDatabaseName is the default OpenTSDB database if none is specified
+	DefaultDatabaseName = "opentsdb"
+)
+
+// SeriesWriter defines the interface for the destination of the data.
+type SeriesWriter interface {
+	WriteSeries(database, retentionPolicy string, points []influxdb.Point) (uint64, error)
+}
+
+// An InfluxDB input class to accept OpenTSDB's telnet protocol
+// Each telnet command consists of a line of the form:
+//   put sys.cpu.user 1356998400 42.5 host=webserver01 cpu=0
+type Server struct {
+	wrtier SeriesWriter
+
+	database string
+
+	listener *net.TCPListener
+}
+
+func NewServer(w SeriesWriter, db string) *Server {
+	s := &Server{}
+
+	s.wrtier = w
+
+	s.database = db
+
+	return s
+}
+
+func (s *Server) ListenAndServe(listenAddress string) {
+	var err error
+
+	addr, err := net.ResolveTCPAddr("tcp4", listenAddress)
+	if err != nil {
+		log.Println("TSDBServer: ResolveTCPAddr: ", err)
+		return
+	}
+
+	s.listener, err = net.ListenTCP("tcp", addr)
+	if err != nil {
+		log.Println("TSDBServer: Listen: ", err)
+		return
+	}
+
+	defer s.listener.Close()
+	s.HandleListener(s.listener)
+}
+
+func (s *Server) HandleListener(socket *net.TCPListener) {
+	for {
+		// Listen for an incoming connection.
+		conn, err := socket.Accept()
+		if err != nil {
+			log.Println("Error accepting: ", err.Error())
+		}
+		// Handle connections in a new goroutine.
+		go s.HandleConnection(conn)
+	}
+}
+
+func (s *Server) HandleConnection(conn net.Conn) {
+	reader := bufio.NewReader(conn)
+	tp := textproto.NewReader(reader)
+
+	defer conn.Close()
+
+	for {
+		line, err := tp.ReadLine()
+		if err != nil {
+			return
+		}
+
+		inputStrs := strings.Fields(line)
+
+		if len(inputStrs) == 1 && inputStrs[0] == "version" {
+			conn.Write([]byte("InfluxDB TSDB proxy"))
+			continue
+		}
+
+		if len(inputStrs) < 4 || inputStrs[0] != "put" {
+			log.Println("TSDBServer: malformed line, skipping: ", line)
+			continue
+		}
+
+		name := inputStrs[1]
+		tsStr := inputStrs[2]
+		valueStr := inputStrs[3]
+		tagStrs := inputStrs[4:]
+
+		var t time.Time
+		ts, err := strconv.ParseInt(tsStr, 10, 64)
+		if err != nil {
+			log.Println("TSDBServer: malformed timestamp, skipping: ", tsStr)
```

Comment:
This logging is useful, but we do need to think carefully about it. In the event of a stream full of errors, there will be a lot of log messages. Not sure what else to do, but it could be an issue.